### DISCUSSION TITLE
Memory Usage for Simple Scenes

### DISCUSSION THREAD

**Discussion by a-day-old-bagel** [jrouwe/JoltPhysics#727] [Q&A]
After tracking down the reason my program was using close to 300 MB of memory in release mode, I found that Jolt itself is allocating the vast majority of that.

So I went and looked at the Jolt sample app, also compiled in release mode, and found that it, too, uses around 300 MB of memory.

I'm not assuming this usage is is unwarranted, or in error. I just wanted to ask if anyone has any off-the-cuff remarks as to what part of Jolt's functionality requires this much memory, and if there's an easy way to reduce that usage in the case that it isn't actually entirely necessary.

NOTE: This is in the context of Windows, built with VS 2022, specifically. I didn't check any other platforms for corroboration.

**Comment by jrouwe** [comment]
Hello,

I'm not sure how you get to 300 MB, when I run the Samples app, Distribution build, Double Precision mode (more memory than Single) and start the Funnel Test, the entire application uses 85 MB. In Release mode this increases to 105 MB (mainly because it allocates a 2 MB profiling buffer per thread, turn off JPH_PROFILE_ENABLED if you don't want that). This includes a lot of rendering stuff, so a fairer test is to start PerformanceTest. In a Distribution build it looks like this:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/8de1b84b-0b4e-430f-b5fc-cf4d8eda1231)

I've highlighted the 2 allocations that dominate the total memory usage of 48 MB. 

The first is this one:

https://github.com/jrouwe/JoltPhysics/blob/e54a90192265e46193d004d8942f8bd0156c6c74/PerformanceTest/PerformanceTest.cpp#L206

This is a temporary allocator. If you have some memory lying around that you can spare during ```PhysicsSystem::Update``` then give it to this allocator. The buffer given here is 32 MB because the performance test simulates a huge pile with many contacts. For normal applications 10 MB should be plenty.

The second one is this:

https://github.com/jrouwe/JoltPhysics/blob/e54a90192265e46193d004d8942f8bd0156c6c74/PerformanceTest/PerformanceTest.cpp#L265

10240, 65536, and 20480 determine the max number of bodies, max number of body pairs and max number of constraints. Again these numbers are much higher than what a normal application uses. E.g. reducing the max number of constraints from 20480 to 10240 frees up 3 MB. Halving the max number of body pairs from 65536 to 32768 frees up another 3 MB etc.

So for a normal scene, Jolt should use < 20 MB. Obviously, if you start creating Shapes like MeshShapes or HeightFieldShapes, the memory usage increases. One of the most complex scenes in the Samples app is the RigPileTest which adds another 12 MB for the scene:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/c107f5a6-c85b-4bfd-a7bd-c7363855161c)

**Comment by a-day-old-bagel** [comment]
Thanks for taking the time to lay that all out for me. I'll try to reproduce what you have there, as a sort of sanity check.

So far, merely building for distribution mode has not gotten me down that far, but I might have some other environmental issue setting flags incorrectly or something.
